KAFKA-9615: Clean up task/producer create and close

vvcephei opened the pull request:
Consolidates task/producer management.

Now, exactly one component manages the creation and destruction of Producers,
whether they are per-thread or per-task.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
wip

Commit:
step 1: pull out task creators

Commit:
step 2: encapsulate producers

Commit:
step 3: clean up StreamsProducer

Commit:
step 4: clean up abstract task creator

Commit:
final cleanup

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -0,0 +1,237 @@
+
+import static org.apache.kafka.streams.StreamsConfig.EXACTLY_ONCE;
+
+class ActiveTaskCreator {
```
Pulled out of StreamThread (could have been done a long time ago, since it was a static class anyway). I didn't embed it in TaskManager, just to keep the file size lower.

I also dropped AbstractTaskCreator, since the creation of Active and Standby tasks are only similar, not exactly the same. We weren't really using the abstraction for much except de-duplicating a few field declarations. On the con side, the abstraction made it hard to see that we were requiring several arguments for Standby task creation that were actually not ever used. The indirection also made it harder to read the task creation logic.

abbccdda replied:
Sounds good to me

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollector.java:
```diff
@@ -25,7 +25,7 @@
 
 import java.util.Map;
 
-public interface RecordCollector extends AutoCloseable {
```
We never used it as an AutoCloseable, and having it makes it hard to trace the callers.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -262,7 +262,7 @@ public void close() {
         if (eosEnabled) {
             streamsProducer.abortTransaction();
         }
-        streamsProducer.close();
```
Only the creator may close the producer, but we can go ahead and call `flush()` here.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -544,11 +306,7 @@ public static StreamThread create(final InternalTopologyBuilder builder,
 
         final ThreadCache cache = new ThreadCache(logContext, cacheSizeBytes, streamsMetrics);
 
-        final Map<TaskId, Producer<byte[], byte[]>> taskProducers = new HashMap<>();
```
Managed fully inside the task factory now.

You'll also notice a bunch of references to the producers are similarly gone in the following lines.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1127,9 +868,7 @@ StreamThread updateThreadMetadata(final String adminClientId) {
             this.state().name(),
             getConsumerClientId(this.getName()),
             getRestoreConsumerClientId(this.getName()),
-            threadProducer == null ?
-                Collections.emptySet() :
-                Collections.singleton(getThreadProducerClientId(this.getName())),
```
Sooo many switch statements are now gone.

guozhangwang replied:
Nice!

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1198,22 +933,7 @@ public String toString(final String indent) {
-            }
-        }
-        return result;
+        return taskManager.producerMetrics();
```
We don't manage the producers anymore, so we can defer to the taskManager (who will defer to the active task creator, but that's none of the thread's business)

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:
```diff
@@ -52,39 +51,20 @@
-            throw new IllegalArgumentException("applicationId and taskId must either be both null or both be not null");
-        }
-
+    public StreamsProducer(final Producer<byte[], byte[]> producer,
+                           final boolean eosEnabled,
+                           final LogContext logContext,
+                           final String applicationId) {
```
This is a bit on the side, but there was some wacky stuff going on in here, and afaict, the only purpose of all the nullable fields was to allow including the task id in exception messages. Do we really need to do that? If so, I'll just add the logContext to the exception message instead, since it already has the task id in it.

guozhangwang replied:
Actually the `logContext.logPrefix()` should have the format of `stream-thread [%s] task [%s]` already, so all the log4j entries are good. For exception messages, we can just get the `prefix` and then encode that into the exception message.

vvcephei replied:
Yep, that's what I was thinking. I'll go ahead and do it.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:
```diff
@@ -141,10 +120,9 @@ private void maybeBeginTransaction() throws ProducerFencedException {
-                    "Error encountered sending record to topic %s%s due to:%n%s",
-                    record.topic(),
-                    taskId == null ? "" : " " + logMessage,
-                    uncaughtException.toString());
```
no need to add the exception to the message, since we also pass it as the cause below.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsProducer.java:
```diff
@@ -212,17 +190,6 @@ public void flush() {
         producer.flush();
     }
 
-    public void close() {
```
No one who has a reference to the StreamsProducer has any business calling close, so the method is gone now.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -227,7 +227,7 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
 
         if (!standbyTasksToCreate.isEmpty()) {
-            standbyTaskCreator.createTasks(mainConsumer, standbyTasksToCreate).forEach(this::addNewTask);
+            standbyTaskCreator.createTasks(standbyTasksToCreate).forEach(this::addNewTask);
```
Standby tasks don't need the mainConsumer (obvious, in retrospect).

vvcephei reviewed (commented):
@guozhangwang , do you mind taking a look at this refactor?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -624,9 +624,11 @@ StandbyTask standbyTask(final TopicPartition partition) {
-    // Instead, we should register and record the metrics properly inside of the record collector.
-    Map<TaskId, StreamTask> fixmeStreamTasks() {
-        return tasks.values().stream().filter(t -> t instanceof StreamTask).map(t -> (StreamTask) t).collect(Collectors.toMap(Task::id, t -> t));
+    Map<MetricName, Metric> producerMetrics() {
```
nit: These two functions are not for testing only.

vvcephei replied:
Thanks for the feedback; I didn't understand this particular comment, though.

guozhangwang replied:
The comment line above these two method declaration says `the following functions are for test only`, but these two functions are not.

vvcephei replied:
Ah, I just found what you were talking about:
```
    // below are for testing only
```

I didn't notice that up there.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -81,9 +81,8 @@
-                final StreamThread.AbstractTaskCreator<? extends Task> activeTaskCreator,
-                final StreamThread.AbstractTaskCreator<? extends Task> standbyTaskCreator,
-                final Map<TaskId, Producer<byte[], byte[]>> taskProducers,
+                final ActiveTaskCreator activeTaskCreator,
```
In `shutdown(final boolean clean)` we should also release task producers as well right?

vvcephei replied:
Yep. Good catch.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:
```diff
@@ -262,7 +262,7 @@ public void close() {
             streamsProducer.abortTransaction();
         }
-        streamsProducer.close();
+        streamsProducer.flush();
```
I'm wondering if we are introducing a latency regression without EOS here: in the old code when closing without EOS we actually do nothing, and now we would block on flushing. 

On the other hand, flushing maybe needed when we close a task to make sure all the tasks' records are acked already.

If the task is in RUNNING before shutting down, we would always commit before closing, so `flush` is already called; if the task is in RESTORING / SUSPENDED there's nothing written from this task, so a `flush` is not needed. So I think it is safe to not call `flush` after all.

mjsax replied:
Sounds reasonable to me

vvcephei replied:
Ok, I'll swap this out.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java:
```diff
@@ -612,25 +582,4 @@ public void shouldFailOnEosAbortTxFatal() {
     }
-
-    @Test
-    public void shouldFailOnCloseFatal() {
```
We do not created new test classes of ActiveTaskCreator / StandbyTaskCreator, but we should still have those coverage to make sure the exception thrown from producers are wrapped correctly. 

Also it seems in the new code we no long rethrow -- is that intentional. I left a comment above.

vvcephei replied:
Good catch. Actually, there's a bunch of coverage missing from TaskManager. I'll add several more tests.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -0,0 +1,237 @@
+            try {
+                threadProducer.close();
+            } catch (final RuntimeException e) {
+                log.error("Failed to close producer due to the following error:", e);
```
The old code we re-throw exceptions whereas here we just swallow the error. Is that intentional?

vvcephei replied:
Ah, thanks for noticing this. I meant to ask about it in here.

I thought it was strange that we would only re-throw when closing the task producers, not the thread producer. It seems like we should do the same thing in both cases, but which thing should we do?

I went with an error log in both cases, but it sounds like you wanted to throw the exception instead. Should we also rethrow for the thread producer?

guozhangwang replied:
Yeah I think we should make them consistent: previously the closing producer is within `task#close` and if it is called via `closeDirty` we should make sure it never throws. Now since it is extracted out of the `close` call we should just rethrow for both cases.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/state/KeyValueStoreTestDriver.java:
```diff
@@ -201,7 +201,7 @@ private KeyValueStoreTestDriver(final StateSerdes<K, V> serdes) {
             new TaskId(0, 0),
             consumer,
-            new StreamsProducer(logContext, producer),
+            new StreamsProducer(producer, null != null, logContext, null),
```
`null != null`?

abbccdda replied:
lol

vvcephei replied:
Oops! Auto-refactoring. I already made two passes to clean these up, looks like I missed one.

guozhangwang reviewed (commented):
I made a pass over the PR.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -0,0 +1,237 @@
+        return createdTasks;
+    }
+
+    public void releaseProducer() {
```
`closeThreadProducer()`

The method below is not public -- does this one need to be public?

abbccdda replied:
even better `maybeCloseThreadProducer`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -0,0 +1,237 @@
+        }
+    }
+
+    void releaseProducer(final TaskId id) {
```
`closeProducerForTask(TaskId)` ?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -0,0 +1,237 @@
+        }
+    }
+
+    public InternalTopologyBuilder builder() {
```
Does this need to be public?

vvcephei replied:
Hah, doesn't need to be there at all, actually.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -0,0 +1,237 @@
+        return builder;
+    }
+
+    public StateDirectory stateDirectory() {
```
Does this need to be public?

mjsax reviewed (commented):
Nice cleanup. There is quite some overlap with https://github.com/apache/kafka/pull/8215 even if both PR address a different issue.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -0,0 +1,237 @@
+
+            if (threadProducer == null) {
+                // create one producer per task for EOS
+                // TODO: after KIP-447 this would be removed
```
I don't think we need to keep this TODO, as only after a stream 3.0 is there, we shall remove the support for task producer.

abbccdda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java:
```diff
@@ -0,0 +1,237 @@
+                log.error("Failed to close producer due to the following error:", e);
+            }
+        }
+        if (!taskProducers.isEmpty()) {
```
Should we throw illegal state first, since we are already in an error state?

vvcephei replied:
If we're going to rename the method to specify that it should do exactly "close thread producer", then this check is no longer appropriate.

Commit:
format exception messages in StreamsProducer

Commit:
misc cr

Commit:
fixed missing coverage in TaskManagerTest

Commit:
fix style

Commit:
repair tests

Commit:
catch producer close exceptions during unclean shutdown

vvcephei commented:
Hey @guozhangwang , @mjsax , and @abbccdda ,

I've addressed all the feedback. In particular I added about 500 lines of missing tests to TaskManagerTest. It looks like we've done a fair amount of un-unit-tested refactoring in TaskManagerTest, so I figured I'd go ahead and add the missing coverage while I was adding coverage for this refactor itself. Be forewarned, the tests are... extensive.

Commit:
suppress complexity

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -345,7 +359,11 @@ void handleLostAll() {
+                try {
+                    activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(task.id());
+                } catch (final RuntimeException e) {
+                    log.debug("Error handling lostAll", e);
```
1. Let's make it a warn instead of a debug.
2. The error message can be more specific here: `Error closing task producer for task {} while handling lostAll`.

abbccdda replied:
+1

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -419,14 +439,33 @@ void shutdown(final boolean clean) {
+                    if (clean) {
+                        firstException.compareAndSet(null, e);
+                    } else {
+                        log.warn("Ignoring an exception while closing task producer.", e);
```
Ditto here about error message

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -440,6 +872,41 @@ public void shouldCommitActiveAndStandbyTasks() {
     }
 
+    @Test
+    public void shouldNotCommitActiveAndStandbyTasks() {
```
nit: ...WhileRebalanceInProgress

guozhangwang reviewed (approved):
@vvcephei I made a pass over the testing code, and only have a couple minor comments.

After addressing them please feel free to merge.

Commit:
minor final CR feedback

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsProducerTest.java:
```diff
@@ -254,30 +224,30 @@ public void shouldNotCloseProducerIfEosDisabled() {
-        assertTrue(eosMockProducer.history().isEmpty());
-        assertThat(eosMockProducer.uncommittedRecords().size(), equalTo(1));
-        assertThat(eosMockProducer.uncommittedRecords().get(0), equalTo(record));
+        assertThat(eosMockProducer.transactionInFlight(), is(true));
```
I was wondering what's the standard for using `assertThat` vs `assertTrue`? Do we have a convention to follow?

vvcephei replied:
`assertThat` is nicer in general, but it doesn't really matter. In this case, IDEA offered to translate, and I was already changing a lot of assertions, so I just accepted the translation.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -24,14 +24,20 @@
 import org.apache.kafka.clients.consumer.Consumer;
 import org.apache.kafka.clients.consumer.ConsumerRecord;
 import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.Metric;
```
Are the tests added in TaskManager only trying for more coverage? @vvcephei

vvcephei replied:
yep, that's right.

abbccdda reviewed (approved):
LGTM, only have some minor comments about tests. After addressing them, feel free to merge.

vvcephei commented:
The last commit was trivial, and the Streams tests passed locally for me, so I'm going to go ahead and merge.
